main/src/hadd.cxx

```diff
@@ -90,281 +99,382 @@
-            if (argv[a][3] != '\0') {
-               std::cerr << "Error: option -ff should not have any suffix: " << argv[a] << " (suffix has been ignored)\n";
-            }
-         }
-         char ft[7];
-         for (int alg = 0; !useFirstInputCompression && alg <= 5; ++alg) {
-            for( int j=0; j<=9; ++j ) {
-               const int comp = (alg*100)+j;
-               snprintf(ft,7,"-f%s%d",prefix,comp);
-               if (!strcmp(argv[a],ft)) {
-                  farg = true;
-                  force = kTRUE;
-                  newcomp = comp;
-                  break;
-               }
-            }
-         }
-         if (!farg) {
-            // Bad argument
-            std::cerr << "Error: option " << argv[a] << " is not a supported option.\n";
+            std::cerr << "[err] Invalid flag: " << arg << "\n";
+            return EFlagResult::kError;
          }
-         ++ffirst;
-      } else if (!outputPlace) {
-         outputPlace = a;
       }
+      ++cur;
+   }
+
+   return EFlagResult::kParsed;
+}
+
+// Returns nullopt if any of the flags failed to parse.
+// If an unknown flag is encountered, it will print a warning and go on.
+static std::optional<HAddArgs> ParseArgs(int argc, char **argv)
+{
+   HAddArgs args{};
+
+   for (int argIdx = 1; argIdx < argc; ++argIdx) {
+      const char *argRaw = argv[argIdx];
+      if (!*argRaw) continue;
+      if (argRaw[0] == '-' && argRaw[1] != '\0') {
+         // parse flag
+         const char *arg = argRaw + 1;
+         bool validFlag = false;
+
+#define PARSE_FLAG(func, ...)                    \
+   do {                                          \
+      if (!validFlag) {                          \
+         const auto res = func(__VA_ARGS__);     \
+         if (res == EFlagResult::kError)          \
+            return {};                           \
+         validFlag = res == EFlagResult::kParsed; \
+      }                                          \
+   } while (0)
+
+         PARSE_FLAG(FlagToggle, arg, "T", args.fNoTrees);
+         PARSE_FLAG(FlagToggle, arg, "a", args.fAppend);
+         PARSE_FLAG(FlagToggle, arg, "k", args.fSkipErrors);
+         PARSE_FLAG(FlagToggle, arg, "O", args.fReoptimize);
+         PARSE_FLAG(FlagToggle, arg, "dbg", args.fDebug);
+         PARSE_FLAG(FlagArg, argc, argv, argIdx, "d", args.fWorkingDir);
+         PARSE_FLAG(FlagArg, argc, argv, argIdx, "j", args.fNProcesses, ConvertNProcesses);
+         PARSE_FLAG(FlagArg, argc, argv, argIdx, "cachesize", args.fCacheSize, ConvertCacheSize);
+         PARSE_FLAG(FlagArg, argc, argv, argIdx, "experimental-io-features", args.fFeatures);
+         PARSE_FLAG(FlagArg, argc, argv, argIdx, "n", args.fMaxOpenedFiles);
+         PARSE_FLAG(FlagArg, argc, argv, argIdx, "v", args.fVerbosity);
+         PARSE_FLAG(FlagF, arg, args);
+
+#undef PARSE_FLAG
+
+         if (!validFlag)
+            std::cerr << "[warn] Unknown flag: " << argRaw << "\n";
```

Comment:
Similar comment for warning, we should decide on a single way to raise all warnings